backend/drm/drm.c

```diff
@@ -1613,3 +1624,141 @@ void destroy_drm_connector(struct wlr_drm_connector *conn) {
 	wl_list_remove(&conn->link);
 	free(conn);
 }
+
+int wlr_drm_backend_get_read_only_fd(struct wlr_backend *backend) {
+	assert(backend);
+
+	struct wlr_drm_backend *drm = get_drm_backend_from_backend(backend);
+	char *path = drmGetDeviceNameFromFd2(drm->fd);
+	if (!path) {
+		wlr_log(WLR_ERROR, "Failed to get device name from DRM fd");
+		return -1;
+	}
+
+	int fd = open(path, O_RDWR);
+	if (fd < 0) {
+		wlr_log_errno(WLR_ERROR, "Unable to clone DRM fd for client fd");
+		free(path);
+		return -1;
+	}
+
+	if (drmIsMaster(fd) && drmDropMaster(fd) < 0) {
+		wlr_log_errno(WLR_ERROR, "Failed to drop master");
+		return -1;
+	}
+
+	return fd;
+}
+
+int wlr_drm_create_lease(struct wlr_output **outputs, size_t n_outputs,
+		uint32_t *lessee_id) {
+	assert(outputs);
+
+	if (n_outputs == 0) {
+		wlr_log(WLR_ERROR, "Can't lease 0 outputs");
+		return -1;
+	}
+
+	struct wlr_drm_backend *drm =
+			get_drm_backend_from_backend(outputs[0]->backend);
+
+	int n_objects = 0;
+	for (size_t i = 0; i < n_outputs; ++i) {
+		struct wlr_drm_connector *conn =
+				get_drm_connector_from_output(outputs[i]);
+		assert(conn->lessee_id == 0);
+
+		if (conn->backend != drm) {
+			wlr_log(WLR_ERROR, "Can't lease output from different backends");
+			return -1;
+		}
+
+		/* conn->id + conn->crtc->id + conn->crtc->primary->id */
+		n_objects += 1;
+		if (conn->crtc) {
```

Comment:
We shouldn't create a lease with no CRTC for a connector. Clients won't be able to do anything with the connectors if there's no CRTC.